File: ios/Kilkari/Screens/HealthScreen.swift
import SwiftUI
import SwiftData
import KilkariCore

/// The health hub: where the vaccinations stand, and the five things underneath them.
struct HealthScreen: View {
    let baby: Baby

    @Environment(\.accent) private var accent
    @Query private var doses: [VaccineDose]
    @Query(sort: \GrowthRecord.date) private var growth: [GrowthRecord]
    @Query(sort: \Appointment.startAt) private var appointments: [Appointment]
    @Query(sort: \LogEntry.startAt, order: .reverse) private var entries: [LogEntry]

    private var groups: [VaccineGroupState] {
        VaccinePlan.groups(for: baby, doses: doses)
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 12) {
                Text("Health").font(KFont.screenTitle).foregroundStyle(KC.ink)

                NavigationLink(value: Route.vaccines) { vaccineHero }
                    .buttonStyle(SpringPress())

                // Three hues across the five, a row at a time: green for what grows, coral
                // for what asks something of you, blue for the people you keep on file. Five
                // different colours was a grid that took longer to read than one colour did.
                HStack(alignment: .top, spacing: 10) {
                    tile(Route.growth, "scalemass.fill", KC.leafDeep, KC.leafWash, "Growth", growthDetail)
                    tile(Route.teeth, "mouth.fill", KC.leafDeep, KC.leafWash, "Teeth", teethDetail)
                }
                HStack(alignment: .top, spacing: 10) {
                    tile(Route.meds, "pills.fill", KC.coralDeep, KC.coralWash, "Medications", medsDetail)
                    tile(Route.appointments, "stethoscope", KC.coralDeep, KC.coralWash, "Appointments", apptDetail)
                }
                HStack(alignment: .top, spacing: 10) {
                    tile(Route.doctors, "cross.case.fill", KC.skyDeep, KC.skyWash, "Doctors", "Add the people you see")
                    Color.clear.frame(maxWidth: .infinity)
                }

                SectionLabel("Recent").padding(.top, 4)
                recent
            }
            .padding(.horizontal, 16)
            .padding(.top, 18)
            .padding(.bottom, 110)
        }
    }

    private var vaccineHero: some View {
        let done = groups.filter(\.isComplete).count
        let next = VaccinePlan.next(groups)
        return GradientCard(colors: [KC.tealDeep, KC.seaMid]) {
            VStack(alignment: .leading, spacing: 10) {
                HStack {
                    Image(systemName: "syringe.fill").foregroundStyle(.white.opacity(0.9))
                    Text("Vaccinations")
                        .font(KFont.display(20, .bold))
                        .foregroundStyle(.white)
                    Spacer()
                    Text("IAP")
                        .font(KFont.sans(11, .semibold))
                        .foregroundStyle(.white)
                        .padding(.horizontal, 8).padding(.vertical, 3)
                        .background(.white.opacity(0.2)).clipShape(Capsule())
                }
                Text(next.map { "\(done) given · next \($0.def.label) \($0.dueText)" }
                     ?? "\(done) given · all done")
                    .font(KFont.sans(13))
                    .foregroundStyle(.white.opacity(0.9))

                GeometryReader { geo in
                    ZStack(alignment: .leading) {
                        Capsule().fill(.white.opacity(0.25))
                        Capsule().fill(.white)
                            .frame(width: geo.size.width * progress)
                    }
                }
                .frame(height: 6)
            }
        }
    }

    private var progress: Double {
        let total = groups.reduce(0) { $0 + $1.total }
        guard total > 0 else { return 0 }
        return Double(groups.reduce(0) { $0 + $1.givenCount }) / Double(total)
    }

    private func tile(
        _ route: Route, _ symbol: String, _ fg: Color, _ wash: Color,
        _ title: String, _ detail: String
    ) -> some View {
        NavigationLink(value: route) {
            VStack(alignment: .leading, spacing: 12) {
                Circle().fill(.white.opacity(0.85)).frame(width: 44, height: 44)
                    .overlay {
                        Image(systemName: symbol)
                            .font(.system(size: 20, weight: .semibold))
                            .foregroundStyle(fg)
                    }
                VStack(alignment: .leading, spacing: 2) {
                    Text(title).font(KFont.sans(16, .bold)).foregroundStyle(KC.ink)
                    Text(detail)
                        .font(KFont.sans(12)).foregroundStyle(KC.mutedStrong)
                        .lineLimit(2).multilineTextAlignment(.leading)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(16)
            .frame(minHeight: 104, alignment: .topLeading)
            .background(wash)
            .clipShape(RoundedRectangle(cornerRadius: 26, style: .continuous))
            .clay(corner: 26)
        }
        .buttonStyle(SpringPress())
    }

    private var growthDetail: String {
        guard let last = growth.last else { return "No measurements yet" }
        return [last.weightKg.map { String(format: "%.1f kg", $0) },
                last.lengthCm.map { String(format: "%.1f cm", $0) }]
            .compactMap { $0 }.joined(separator: " · ")
    }

    private var teethDetail: String {
        let count = entries.filter { $0.kind == .tooth }.count
        return "\(count) of 20 through"
    }

    private var medsDetail: String {
        let last = entries.first { $0.kind == .medicine }
        return last.map { $0.note ?? "Recorded" } ?? "None active"
    }

    private var apptDetail: String {
        guard let next = appointments.first(where: { $0.startAt >= .now }) else { return "Nothing booked" }
        return "Next: \(Fmt.date(next.startAt)) \(Fmt.time(next.startAt))"
    }

    private var recent: some View {
        let rows = growth.suffix(3).reversed().map {
            ($0.weightKg.map { w in String(format: "Weight %.1f kg", w) } ?? "Measurement", $0.date)
        }
        return KCard {
            if rows.isEmpty {
                Text("Nothing recorded yet.")
                    .font(KFont.sans(13)).foregroundStyle(KC.muted).padding(14)
            } else {
                VStack(spacing: 0) {
                    ForEach(Array(rows.enumerated()), id: \.offset) { i, row in
                        HStack {
                            IconBadge(symbol: "scalemass.fill", tint: KC.leafDeep,
                                      background: KC.leafBg, size: 32, corner: 10, iconSize: 15)
                            Text(row.0).font(KFont.sans(14)).foregroundStyle(KC.ink)
                            Spacer()
                            Text(Fmt.date(row.1)).font(KFont.sans(12)).foregroundStyle(KC.muted)
                        }
                        .padding(.horizontal, 14).padding(.vertical, 11)
                        if i != rows.count - 1 { Rectangle().fill(KC.divider).frame(height: 1) }
                    }
                }
            }
        }
    }
}

File: android/app/src/main/java/com/kilkari/ui/screens/HealthScreen.kt
package com.kilkari.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.kilkari.data.seed.VaccineSchedules
import com.kilkari.domain.Fmt
import com.kilkari.domain.ToothChart
import com.kilkari.ui.KilkariViewModel
import com.kilkari.ui.components.GradientCard
import androidx.compose.foundation.shape.CircleShape
import com.kilkari.ui.components.KCard
import com.kilkari.ui.components.KIcons
import com.kilkari.ui.components.SectionLabel
import com.kilkari.ui.nav.NavActions
import com.kilkari.ui.nav.Routes
import com.kilkari.ui.theme.KDepth
import com.kilkari.ui.theme.Display
import com.kilkari.ui.theme.KC
import com.kilkari.ui.theme.Sans
import com.kilkari.ui.theme.ScreenTitle
import java.time.LocalDateTime

/** Hub for everything clinical: vaccines hero, four tiles, and a recent-activity list. */
@Composable
fun HealthScreen(vm: KilkariViewModel, go: NavActions) {
    val settings by vm.settings.collectAsStateWithLifecycle()
    val groups by vm.vaccineGroups.collectAsStateWithLifecycle()
    val nextVac by vm.nextVaccine.collectAsStateWithLifecycle()
    val growth by vm.growth.collectAsStateWithLifecycle()
    val teeth by vm.teeth.collectAsStateWithLifecycle()
    val meds by vm.medications.collectAsStateWithLifecycle()
    val appointments by vm.appointments.collectAsStateWithLifecycle()
    val timeline by vm.timeline.collectAsStateWithLifecycle()
    val doctors by vm.doctors.collectAsStateWithLifecycle()

    val doneCount = groups.sumOf { it.doneCount }
    val total = groups.sumOf { it.count }
    val pct = if (total == 0) 0 else doneCount * 100 / total
    val latestGrowth = growth.lastOrNull()
    val activeMeds = meds.filter { it.active }
    val nextAppt = appointments.firstOrNull { !it.startAt.isBefore(LocalDateTime.now()) }

    Column(
        Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 16.dp)
            .padding(top = 18.dp, bottom = KDepth.navClearance),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Text("Health", style = ScreenTitle, color = KC.Ink)

        GradientCard(listOf(KC.TealDeep, KC.SeaMid), onClick = { go.push(Routes.VACCINES) }) {
            Row(
                Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(KIcons["vaccines"], null, tint = Color.White, modifier = Modifier.size(26.dp))
                Box(
                    Modifier
                        .clip(RoundedCornerShape(999.dp))
                        .background(Color.White.copy(alpha = 0.2f))
                        .padding(horizontal = 9.dp, vertical = 3.dp),
                ) {
                    Text(
                        VaccineSchedules.byId(settings.scheduleId).shortName,
                        fontFamily = Sans, fontSize = 12.sp, color = Color.White,
                    )
                }
            }
            Text(
                "Vaccinations",
                fontFamily = Display, fontWeight = FontWeight.Bold, fontSize = 18.sp, color = Color.White,
            )
            Text(
                nextVac?.let { "$doneCount given · next ${it.label} on ${Fmt.date(it.dueDate)}" }
                    ?: "$doneCount given",
                fontFamily = Sans, fontSize = 13.sp, color = Color.White.copy(alpha = 0.9f),
                maxLines = 2, overflow = TextOverflow.Ellipsis,
            )
            Box(
                Modifier
                    .fillMaxWidth()
                    .height(6.dp)
                    .clip(RoundedCornerShape(3.dp))
                    .background(Color.White.copy(alpha = 0.25f)),
            ) {
                Box(
                    Modifier
                        .fillMaxWidth(pct / 100f)
                        .height(6.dp)
                        .clip(RoundedCornerShape(3.dp))
                        .background(Color.White),
                )
            }
        }

        Row(
            Modifier.fillMaxWidth().height(IntrinsicSize.Min),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            HealthTile(
                "monitor_weight", KC.LeafDeep, KC.LeafWash, "Growth",
                latestGrowth?.let {
                    listOfNotNull(
                        it.weightKg?.let { w -> Fmt.weight(w, settings.metricUnits) },
                        it.lengthCm?.let { l -> Fmt.length(l, settings.metricUnits) },
                        it.headCm?.let { h -> Fmt.length(h, settings.metricUnits) },
                    ).joinToString(" · ")
                }?.ifBlank { "No measurements yet" } ?: "No measurements yet",
                Modifier.weight(1f).fillMaxHeight(),
            ) { go.push(Routes.GROWTH) }
            HealthTile(
                "dentistry", KC.LeafDeep, KC.LeafWash, "Teeth",
                // Points at the tooth actually due next rather than repeating the first one
                // forever, which stopped being true the moment it came through.
                ToothChart.nextExpected(teeth).let { next ->
                    if (next == null) "All ${ToothChart.TOTAL} through"
                    else "${teeth.size} of ${ToothChart.TOTAL} · next ~${next.second.fromMonth} mo"
                },
                Modifier.weight(1f).fillMaxHeight(),
            ) { go.push(Routes.TEETH) }
        }
        Row(
            Modifier.fillMaxWidth().height(IntrinsicSize.Min),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            HealthTile(
                "pill", KC.CoralDeep, KC.CoralWash, "Medications",
                if (activeMeds.isEmpty()) "None active"
                else "${activeMeds.size} active · ${activeMeds.first().name}",
                Modifier.weight(1f).fillMaxHeight(),
            ) { go.push(Routes.MEDS) }
            HealthTile(
                "stethoscope", KC.CoralDeep, KC.CoralWash, "Appointments",
                nextAppt?.let { "Next: ${Fmt.dayAndDate(it.startAt.toLocalDate())} ${Fmt.time(it.startAt)}" }
                    ?: "Nothing booked",
                Modifier.weight(1f).fillMaxHeight(),
            ) { go.push(Routes.APPOINTMENTS) }
        }
        Row(
            Modifier.fillMaxWidth().height(IntrinsicSize.Min),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            HealthTile(
                "medical_services", KC.SkyDeep, KC.SkyWash, "Doctors",
                if (doctors.isEmpty()) "Add the people you see"
                else "${doctors.size} saved · ${doctors.first().name}",
                Modifier.weight(1f).fillMaxHeight(),
            ) { go.push(Routes.DOCTORS) }
            Box(Modifier.weight(1f))
        }

        SectionLabel("Recent", Modifier.padding(top = 4.dp))
        KCard {
            val recent = buildList {
                growth.takeLast(2).reversed().forEach { g ->
                    val prev = growth.getOrNull(growth.indexOf(g) - 1)?.weightKg
                    val delta = if (prev != null && g.weightKg != null) {
                        " (${Fmt.grams(g.weightKg - prev, settings.metricUnits)})"
                    } else ""
                    add(
                        Triple(
                            "monitor_weight" to KC.LeafDeep,
                            "Weight ${Fmt.weight(g.weightKg, settings.metricUnits)}$delta",
                            g.date,
                        )
                    )
                }
                meds.take(2).forEach { m ->
                    add(Triple("pill" to KC.ClayDeep, "${m.name} started, ${m.dose}", m.startDate))
                }
                timeline.filter { it.icon == "vaccines" }.take(2).forEach { t ->
                    add(Triple("vaccines" to KC.TealDeep, t.title, t.date))
                }
            }.sortedByDescending { it.third }.take(4)

            if (recent.isEmpty()) {
                Text(
                    "Nothing recorded yet.",
                    modifier = Modifier.padding(14.dp),
                    fontFamily = Sans, fontSize = 13.sp, color = KC.Muted,
                )
            }
            recent.forEachIndexed { i, (iconAndTint, text, date) ->
                Row(
                    Modifier.fillMaxWidth().padding(horizontal = 14.dp, vertical = 12.dp),
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Icon(
                        KIcons[iconAndTint.first], null,
                        tint = iconAndTint.second, modifier = Modifier.size(20.dp),
                    )
                    Text(
                        text, modifier = Modifier.weight(1f),
                        fontFamily = Sans, fontSize = 14.sp, color = KC.Ink,
                        maxLines = 2, overflow = TextOverflow.Ellipsis,
                    )
                    Text(Fmt.date(date), fontFamily = Sans, fontSize = 12.sp, color = KC.Muted)
                }
                if (i != recent.lastIndex) {
                    Box(Modifier.fillMaxWidth().height(1.dp).background(KC.Divider))
                }
            }
        }
    }
}

/**
 * One of the five things health is made of, drawn the way the log tiles are: the card itself
 * wears the colour and the icon sits in a white disc on top of it. Five white cards with five
 * small coloured dots is a settings list; five coloured cards is a place.
 *
 * Three hues across the five, a row at a time: green for what grows, coral for what asks
 * something of you, blue for the people you keep on file. Five different colours was a grid
 * that took longer to read than one colour did.
 */
@Composable
private fun HealthTile(
    icon: String,
    fg: Color,
    wash: Color,
    title: String,
    subtitle: String,
    modifier: Modifier = Modifier,
    onClick: () -> Unit,
) {
    KCard(modifier, corner = 26, background = wash, border = null, onClick = onClick) {
        Column(
            Modifier.fillMaxHeight().padding(16.dp).heightIn(min = 104.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Box(
                Modifier
                    .size(44.dp)
                    .clip(CircleShape)
                    .background(Color.White.copy(alpha = 0.85f)),
                contentAlignment = Alignment.Center,
            ) {
                Icon(KIcons[icon], null, tint = fg, modifier = Modifier.size(23.dp))
            }
            Column {
                Text(title, fontFamily = Sans, fontWeight = FontWeight.Bold, fontSize = 16.sp, color = KC.Ink)
                Text(
                    subtitle,
                    fontFamily = Sans, fontSize = 12.sp, color = KC.MutedStrong,
                    modifier = Modifier.padding(top = 2.dp),
                    maxLines = 2, overflow = TextOverflow.Ellipsis,
                )
            }
        }
    }
}
